Gems/Terrain/Assets/Shaders/Terrain/TerrainDetailHelpers.azsli

```diff
@@ -225,26 +252,112 @@ void GetDebugDetailSurface(inout DetailSurface surface, uint material1, uint mat
     surface.m_height = 0.5;
 }
 
-bool GetDetailSurface(inout DetailSurface surface, float2 idUv, float2 uv)
+//Blend a single detail material sample (with two possible material ids) onto a DetailSurface.
+void BlendDetailMaterial(inout DetailSurface surface, uint material1, uint material2, float blend, float2 detailUv, float weight)
 {
-    uint4 material1 = TerrainSrg::m_detailMaterialIdImage.GatherRed(TerrainSrg::DetailSampler, idUv, 0).xyzw;
-    uint4 material2 = TerrainSrg::m_detailMaterialIdImage.GatherGreen(TerrainSrg::DetailSampler, idUv, 0).xyzw;
+    DetailSurface tempSurface;
+    GetDetailSurfaceForMaterial(tempSurface, material1, detailUv);
+    if (material2 != 0xFF)
+    {
+        WeightDetailSurface(tempSurface, weight * blend);
+        AddDetailSurface(surface, tempSurface);
 
-    const float maxBlendAmount = 0xFF;
-    // convert integer of 0-255 to float of 0-1.
-    float4 blends = float4(TerrainSrg::m_detailMaterialIdImage.GatherBlue(TerrainSrg::DetailSampler, idUv, 0).xyzw) / maxBlendAmount;
+        GetDetailSurfaceForMaterial(tempSurface, material2, detailUv);
+        WeightDetailSurface(tempSurface, weight * (1.0 - blend));
+        AddDetailSurface(surface, tempSurface);
+    }
+    else
+    {
+        WeightDetailSurface(tempSurface, weight);
+        AddDetailSurface(surface, tempSurface);
+    }
+}
 
+/*
+Populates a DetailSurface with material data gathered form the 4 nearest samples to detailMaterialIdUv. The weight
+of each detail material's contribution is calculated based on the distance to the center point for that sample (for
+instance, if detailMaterialIdUv falls perfectly inbetween all 4 samples, then each sample will be weighed at 25%).
+Each sample can have two different detail materials defined with a blend value to determine their relative contribution.
+The detailUv is used for samping the textures of each detail material.
+*/
+bool GetDetailSurface(inout DetailSurface surface, float2 detailMaterialIdUv, float2 detailUv)
+{
+    float2 textureSize;
+    TerrainSrg::m_detailMaterialIdImage.GetDimensions(textureSize.x, textureSize.y);
+
+    float2 detailMaterialIdCoord = detailMaterialIdUv * textureSize + 0.5;
+    float2 detailMaterailIdUv = detailMaterialIdCoord;
+    float2 gatherWeight = frac(detailMaterialIdCoord);
+
+    // Adjust the gather weight for better interpolation. This helps avoid diamond-shaped artifacts in binlinear filtering.
+    gatherWeight = 3.0 * gatherWeight * gatherWeight - 2.0 * gatherWeight * gatherWeight * gatherWeight;
+
+    // Using Load() to gather the nearest 4 samples (Gather4() isn't used because of precision issues with uvs).
+    uint4 s1 = TerrainSrg::m_detailMaterialIdImage.Load(int3((detailMaterialIdCoord + int2(0, 1) + textureSize) % textureSize, 0));
```

Comment:
You're doing + textureSize % textureSize to handle negative detailMaterialIdCoord, plus to wrap around at the border of the detailId texture, correct? Maybe add a comment to that effect to explain why you're doing that here.

You're also doing detailMaterialIdCoord + textureSize four times. The int2 is a bit misleading, since that should get implicitly cast to a float during the addition. Also, you're doing % four times, calculating four float2's. But you really only need 4 values. I think you can factor this out to something like

// detailMaterialIdCoord might be negative, so add textureSize to ensure it is positive
float2 positiveDetailMaterialIdCoord = detailMaterialIdCoord + textureSize;
int2 loadCoord0 = int2(positiveDetailMaterialIdCoord % textureSize);
int2 loadCoord1 = int2((positiveDetailMaterialIdCoord + float2(1, 1)) % textureSize);
uint4 s1 = TerrainSrg::m_detailMaterialIdImage.Load(int3(loadCoord0.x, loadCoord1.y, 0));
uint4 s2 = TerrainSrg::m_detailMaterialIdImage.Load(int3(loadCoord1.x, loadCoord1.y, 0));
uint4 s3 = TerrainSrg::m_detailMaterialIdImage.Load(int3(loadCoord1.x, loadCoord0.y, 0));
uint4 s4 = TerrainSrg::m_detailMaterialIdImage.Load(int3(loadCoord0.x, loadCoord0.y, 0));

I'm also not convinced the logic is correct. It seems like this is always going to get the texels down and to the right of detailMaterialIdCoord, which is not what you want if detailMaterialIdCoord is in the top half or left half of the texel. It seems like you could work around this by calculating detailMatrialCoordId as

float2 detailMaterialIdCoord = detailMaterialIdUv * textureSize - 0.5;

instead of + 0.5. That way you always pick the correct 4 texels. You'd just have to make sure the weights stay associated with the correct texels, which should also be doable.

Reply:
Good catch on the float int conversion. I also like calculating the two coordinates for top left and bottom right. I think the 0.5 was needed for part of the logic when I was doing gathers but with Load() I think you're right that it was just offsetting everything by 0.5 now. It's hard to tell for sure because the outcome looks pretty equivalent and seems to line up with the terrain either way. We'll need to build some good tests later that use specifically designed gradients to make sure the heights and detail textures all line up exactly where they should because it's easy to be off by just a little bit and not know it.